### ISSUE TITLE
Popup closes when map is panned or zoomed

### ISSUE THREAD

**Issue by perfectly-preserved-pie** [perfectly-preserved-pie/larentals#86]
Since using `dl.GeoJSON`, if a popup is open and the map is panned or zoomed even a tiny bit, the popup closes. Clicking on a marker also automatically pans to fit the popup inside the screen, but that panning action means the popup closes, so it takes anywhere from 2-4 clicks to get the popup to stay open. Wack behavior.

**Comment by perfectly-preserved-pie** [comment]
https://www.youtube.com/watch?v=n7HUBNmXB5Y

lmao should I just go back to using Folium for the mapping library?

**Comment by perfectly-preserved-pie** [comment]
@anaypat5 hey, thanks for that sweet piece of info, I really appreciate it.

I'm close, but setting the autoPan property via assign errors out.
If I set [the title property](https://leafletjs.com/reference.html#marker-title), the title does show up, indicating that the assign function is working correctly:
```
yo = assign("""function(feature, latlng){
  return L.marker(latlng, {title: 'THIS IS A CUSTOM TITLE'});
  }""")

...

return dl.GeoJSON(
    id=str(uuid.uuid4()),
    data=geojson,
    cluster=True,
    zoomToBoundsOnClick=True,
    superClusterOptions={ # https://github.com/mapbox/supercluster#options
      'radius': 160,
      'minZoom': 3,
    },
    options=dict(pointToLayer=yo),

```

`autoPan` isn't a property of `L.marker` but rather `L.popup` so I used that instead:
But I try to set the popup's autoPan property to False:

```
yo = assign("""
  function(feature, latlng){
    return L.popup(
      latlng,
      {autoPan: false}
    );
  }""")
```

I get a lot of `No match for [dashExtensions.default.function63] in the global window object.` errors in the Dash console.

Do you have a working code snippet from your project you'd be willing to share?

**Comment by anaypat5** [comment]
Unfortunately I don't have access to my work computer at the moment so I can't share my code, but if it actually is a `Popup` property then just directly using the reference code like so should work: 

```
window.dash_props = Object.assign({}, window.dash_props, {
    module: {
        on_each_feature: function(feature, layer, context) {
            if (!feature.properties) {
                return
            }
            if (feature.properties.popup) {
                layer.bindPopup(feature.properties.popup, {autoPan: false})
            }
            if (feature.properties.tooltip) {
                // here you can change all leaflet tooltip options
                layer.bindTooltip(feature.properties.tooltip)
            }
        }
    }
});
```

You have to save this as a .js file (I saved mine as window.js but I'm not sure if the file name matters) and put it in an 'assets' folder where your code is, and then you can implement it in your code using 
```
from dash_extensions.javascript import Namespace
...
ns = Namespace("dash_props", "module")
dl.GeoJSON(..., options=dict(onEachFeature=ns("on_each_feature")))
```

**Comment by perfectly-preserved-pie** [comment]
Holy shit that worked!!! Thank you so much! I was even able to resize the popup without resorting to using CSS (which made the popup and the marker mis-aligned)

```window.dash_props = Object.assign({}, window.dash_props, {
    module: {
        on_each_feature: function(feature, layer, context) {
            if (!feature.properties) {
                return
            }
            if (feature.properties.popup) {
                layer.bindPopup(feature.properties.popup, {
                    // Here you can customize the popup
                    // https://leafletjs.com/reference.html#popup-option
                    autoPan: false,
                    closeButton: false,
                    // Set the maxHeight to 500px if the device is mobile
                    maxHeight: window.innerWidth < 768 ? 500 : null,
                    // Set the maxWidth to 175px if the device is mobile
                    maxWidth: window.innerWidth < 768 ? 175 : null,
                    
                })
            }
        }
    }
});
```

This solves the autopanning issue, which stops the popup from closing when a marker is clicked. Still, that's more of a workaround for my ultimate issue: Do you know if it's possible to stop the popup from closing whenever the map is zoomed or panned? That would completely solve my issue. That behavior wasn't present with `dl.Marker` when I used it.

**Comment by anaypat5** [comment]
Maybe you can try setting `closePopupOnClick=False` in your `dl.Map`, and/or adding `closeOnClick: false` to the popup options you defined in the javascript function. The dash leaflet docs say that `closeOnClick` defaults to what the map property is set as so just the former should work.

**Comment by perfectly-preserved-pie** [comment]
Unfortunately that only modifies the behavior when the map is clicked but not when it's panned/moved :(

**Comment by anaypat5** [comment]
Ah, I was hoping that Leaflet would consider panning as a "click & drag" and it would just work out. I'm currently trying to get a version of your source code working so that I can try my own edits, but in the meantime, I just realized that you said `autoPan` was a `L.popup` property that you tried to set on `L.marker`. Maybe you could try binding the popup to the marker as they would do in Leaflet, and then setting the options on the marker, such as the following:

```
yo = assign("""
  function(feature, latlng){
    return L.popup(latlng).bindPopup(..., {autoPan: false, ...});
  }""")
```

To be honest, if this doesn't error out, I'm not sure how different this would be from what you're already doing with `onEachFeature`, but maybe something under the hood changes here to make this work lol. Although with this I'm not sure if this would still keep the content you currently have set in the popup, but its maybe worth a shot? My Dash Leaflet experience is coming up on like two weeks now haha, so if this doesn't work as intended I'd have to do some more research into it unfortunately.

**Comment by anaypat5** [comment]
Hey, so I took a look at my code and I saw that my popups don't close when the map is panned or zoomed, and the properties I have set in my Javascript function for the popups is `autoClose: false` and `closeOnClick: false`, so try setting both of those in your Javascript function and hopefully that will work.

**Comment by perfectly-preserved-pie** [comment]
I tried that but it didn't change anything. I think the difference might be how we're generating our markers. How are you generating yours? I'm using `dl.GeoJSON` inside a Dash callback like this

```
@callback(
  Output(component_id='lease_geojson', component_property='children'),
  [ <a bunch of inputs>
  ]
...

# Create an empty list for the markers
  markers = []
  # Iterate through the dataframe, create a marker for each row, and append it to the list
  for row in df_filtered.itertuples():
    markers.append(
      dict(
        lat=row.Latitude,
        lon=row.Longitude,
        popup=row.popup_html
        )
    )
  # Generate geojson with a marker for each listing
  geojson = dlx.dicts_to_geojson([{**m} for m in markers])

...

ns = Namespace("dash_props", "module")
  # Generate the map
  return dl.GeoJSON(
    id=str(uuid.uuid4()),
    #children=[dl.Popup(id='2popup')],
    data=geojson,
    cluster=True,
    zoomToBoundsOnClick=True,
    superClusterOptions={ # https://github.com/mapbox/supercluster#options
      'radius': 160,
      'minZoom': 3,
    },
    options=dict(onEachFeature=ns("on_each_feature"))
  )

```

**Comment by perfectly-preserved-pie** [comment]
Good thought - I think it is because you're using a LayerGroup. I was able to stop the popups from closing:

```
# return a LayerGroup with GeoJSON as one of its layers
  return dl.LayerGroup([dl.GeoJSON(
      data=geojson,
      cluster=False,
      superClusterOptions={
          'radius': 160,
          'minZoom': 3,
      },
      zoomToBoundsOnClick=True,
      options=dict(onEachFeature=ns("on_each_feature"))
  )])
```

That stops the popup from closing when the map is panned, but as you can see, without `cluster=True`, I don't have clustering. When I enable clustering, the behavior reverts to closing whenever the map is moved. Are you using clustering in your app? If so, how'd you get around this?

**Comment by perfectly-preserved-pie** [comment]
Looks like this is because of [a limitation of the Supercluster library](https://github.com/mapbox/supercluster/issues/9), but I am curious to know if you found a way around it or are just avoiding `cluster=True`

**Comment by anaypat5** [comment]
Unfortunately the data I am attaching to a GeoJSON is sets of polylines, so I have not really experimented with marker clustering at all. My guess at a work around would be to use a custom `clusterToLayer` function inside the GeoJSON, which might even be as simple as just returning the basic clustering that is already done. You could try modifying the example code [here](https://dash-leaflet.herokuapp.com/#scatter_cluster) to work with your app, but unfortunately I haven't done too much reasearch into clustering, so I hope this leads you in the right direction.

**Comment by perfectly-preserved-pie** [comment]
Resolved via #115 https://github.com/thedirtyfew/dash-leaflet/issues/180#issuecomment-1694490970 🎉
